flink-ml-lib/src/main/java/org/apache/flink/ml/evaluation/binaryeval/BinaryClassificationEvaluator.java

```diff
@@ -0,0 +1,736 @@
+                                            double negativeSum = 0;
+
+                                            for (Tuple4<Double, Long, Boolean, Double> t : values) {
+                                                sum += t.f1;
+                                                cnt++;
+                                                if (t.f2) {
+                                                    positiveSum += t.f3;
+                                                } else {
+                                                    negativeSum += t.f3;
+                                                }
+                                            }
+                                            out.collect(
+                                                    new double[] {
+                                                        1. * sum / cnt * positiveSum,
+                                                        positiveSum,
+                                                        negativeSum
+                                                    });
+                                        })
+                        .returns(double[].class);
+
+        DataStream<Double> areaUnderROC =
+                localAucVariable
+                        .transform(
+                                "reduceInEachPartition",
+                                TypeInformation.of(double[].class),
+                                new CalcAucOperator())
+                        .transform(
+                                "reduceInFinalPartition",
+                                TypeInformation.of(double[].class),
+                                new CalcAucOperator())
+                        .setParallelism(1)
+                        .map(
+                                new MapFunction<double[], Double>() {
+                                    @Override
+                                    public Double map(double[] aucVariable) {
+                                        if (aucVariable[1] > 0 && aucVariable[2] > 0) {
+                                            return (aucVariable[0]
+                                                            - 1.
+                                                                    * aucVariable[1]
+                                                                    * (aucVariable[1] + 1)
+                                                                    / 2)
+                                                    / (aucVariable[1] * aucVariable[2]);
+                                        } else {
+                                            return Double.NaN;
+                                        }
+                                    }
+                                });
+
+        Map<String, DataStream<?>> broadcastMap = new HashMap<>();
+        broadcastMap.put(PARTITION_SUMMARY, partitionSummaries);
+        broadcastMap.put(AREA_UNDER_ROC, areaUnderROC);
```

Comment:
I noticed that `areaUnderROC` is calculated separately from other metrics. Given that all metrics experience similarly to be computed, like a `mapPartition` or `window().apply()`, I have the sense that we may have just one window operator or map partition operator that gets all metrics out. That would greatly simplify the structure of this code. How do you like the idea to try in this way?

Reply:
AreaUnderROC's calculation is different from others. It need keyed by score and do some calculation, other metrics just do some statistics of the input data.